lib/ansible/plugins/inventory/__init__.py

```diff
@@ -444,8 +448,14 @@ def _add_host_to_keyed_groups(self, keys, variables, host, strict=False, fetch_h
                                 new_raw_group_names.append(name)
                         elif isinstance(key, Mapping):
                             for (gname, gval) in key.items():
-                                name = '%s%s%s' % (gname, sep, gval)
-                                new_raw_group_names.append(name)
+                                bare_name = '%s%s%s' % (gname, sep, gval)
+                                if gval is None or gval == '':
+                                    # key's value is empty or undefined
```

Comment:
you mean when the ``key`` from the ``keyed`` is undefined from host_vars ? 
In that case depending on the value of ``strict`` an exception will be raised

Reply:
Yes. None and an empty strings are both 'defined', 'undefined' is does-not-exist. We have a `default` filter that is usually used for undefined (though it can also handle falsey values and empty strings), so the example you posted with an option called `default` made me think it could be confusing if it applied to just empty strings. Since the name isn't the same, you could just remove the mentions of 'undefined' from the documentation and code comments.

I would also remove support for 'None' to limit it to empty strings. Then the behavior for `trailing_separator` would align with `leading_separator`.